File: examples/backends/fastapi/app/routes/refresh.py
"""POST /auth/refresh — Refresh tokens via Cognito."""

import logging

from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse

from .. import ocsf
from ..cognito import cognito_request
from ..config import get_settings
from ..dependencies import destroy_session, require_auth, require_csrf

logger = logging.getLogger(__name__)

router = APIRouter()


@router.post("/auth/refresh")
async def refresh_tokens(
    request: Request,
    _csrf: None = Depends(require_csrf),
    tokens: dict = Depends(require_auth),
):
    if not tokens.get("refresh_token"):
        return JSONResponse({"error": "No refresh token"}, status_code=401)

    s = get_settings()

    try:
        result = await cognito_request(
            "InitiateAuth",
            {
                "AuthFlow": "REFRESH_TOKEN_AUTH",
                "ClientId": s.cognito_client_id,
                "AuthParameters": {
                    "REFRESH_TOKEN": tokens["refresh_token"],
                },
            },
        )
    except Exception as e:
        logger.error("Token refresh error: %s", e)
        email = ocsf._email_from_session(request.state.session)
        ocsf.authentication_event(
            activity_id=ocsf.AuthActivity.SERVICE_TICKET,
            activity_name="Service Ticket",
            status_id=ocsf.Status.FAILURE,
            severity_id=ocsf.Severity.MEDIUM,
            user_email=email,
            message=f"Token refresh failed: {e}",
        )
        destroy_session(request)
        return JSONResponse(
            {"error": "Refresh failed", "message": str(e)}, status_code=401
        )

    auth_result = result.get("AuthenticationResult")
    if not auth_result:
        return JSONResponse({"error": "Refresh failed"}, status_code=500)

    # Update session with new tokens
    request.state.session["tokens"] = {
        "access_token": auth_result["AccessToken"],
        "id_token": auth_result["IdToken"],
        "refresh_token": auth_result.get("RefreshToken", tokens["refresh_token"]),
        "auth_method": tokens.get("auth_method"),
    }

    email = ocsf._email_from_session(request.state.session)
    ocsf.authentication_event(
        activity_id=ocsf.AuthActivity.SERVICE_TICKET,
        activity_name="Service Ticket",
        status_id=ocsf.Status.SUCCESS,
        severity_id=ocsf.Severity.INFORMATIONAL,
        user_email=email,
        message="Token refresh succeeded",
    )

    return {
        "access_token": auth_result["AccessToken"],
        "id_token": auth_result["IdToken"],
        "auth_method": tokens.get("auth_method", "handler"),
    }

File: rust/src/routes/refresh.rs
//! POST /auth/refresh

use axum::extract::State;
use axum::Json;
use std::sync::Arc;

use crate::cognito::client;
use crate::error::AppError;
use crate::ocsf;
use crate::session::middleware::SessionHandle;
use crate::types::{SessionTokens, TokenResponse};

/// Refresh tokens via Cognito, update session.
pub async fn refresh_tokens(
    State(state): State<Arc<crate::AppState>>,
    session: SessionHandle,
) -> Result<Json<TokenResponse>, AppError> {
    let tokens: SessionTokens = {
        let data = session.data.lock().await;
        data.get("tokens")
            .and_then(|v| serde_json::from_value(v.clone()).ok())
            .ok_or(AppError::NotAuthenticated)?
    };

    let refresh_token = tokens
        .refresh_token
        .as_deref()
        .ok_or(AppError::NoRefreshToken)?;

    let email = crate::cognito::jwt::decode_jwt_unverified(&tokens.id_token)
        .ok()
        .and_then(|c| c.email);

    // Call Cognito to refresh
    match client::refresh_tokens(&state.http_client, &state.config, refresh_token).await {
        Ok(result) => {
            let auth_result = result.get("AuthenticationResult");
            let new_access = auth_result
                .and_then(|r| r.get("AccessToken"))
                .and_then(|v| v.as_str());
            let new_id = auth_result
                .and_then(|r| r.get("IdToken"))
                .and_then(|v| v.as_str());

            match (new_access, new_id) {
                (Some(at), Some(it)) => {
                    // Use rotated refresh token if Cognito returned one, else keep existing
                    let new_refresh = auth_result
                        .and_then(|r| r.get("RefreshToken"))
                        .and_then(|v| v.as_str())
                        .map(|rt| Some(rt.to_string()))
                        .unwrap_or_else(|| tokens.refresh_token.clone());

                    let new_tokens = SessionTokens {
                        access_token: at.to_string(),
                        id_token: it.to_string(),
                        refresh_token: new_refresh,
                        auth_method: tokens.auth_method.clone(),
                    };

                    let mut data = session.data.lock().await;
                    data.set("tokens", serde_json::to_value(&new_tokens).unwrap());

                    ocsf::authentication_event(
                        ocsf::ACTIVITY_SERVICE_TICKET,
                        "Service Ticket",
                        ocsf::STATUS_SUCCESS,
                        ocsf::SEVERITY_INFORMATIONAL,
                        email.as_deref(),
                        ocsf::AUTH_PROTOCOL_OAUTH2,
                        "OAuth 2.0/OIDC",
                        "Token refresh succeeded",
                    );

                    Ok(Json(TokenResponse {
                        access_token: new_tokens.access_token,
                        id_token: new_tokens.id_token,
                        auth_method: new_tokens
                            .auth_method
                            .unwrap_or_else(|| "handler".into()),
                    }))
                }
                _ => {
                    // Cognito returned success but no tokens — shouldn't happen
                    Err(AppError::Internal("Refresh failed".into()))
                }
            }
        }
        Err(e) => {
            // Refresh failed — destroy session (Token Handler protocol invariant #6)
            ocsf::authentication_event(
                ocsf::ACTIVITY_SERVICE_TICKET,
                "Service Ticket",
                ocsf::STATUS_FAILURE,
                ocsf::SEVERITY_MEDIUM,
                email.as_deref(),
                ocsf::AUTH_PROTOCOL_OAUTH2,
                "OAuth 2.0/OIDC",
                &format!("Token refresh failed: {}", e),
            );

            let mut destroyed = session.destroyed.lock().await;
            *destroyed = true;
            let mut data = session.data.lock().await;
            data.clear();

            Err(AppError::RefreshFailed(e.to_string()))
        }
    }
}
